presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java

```diff
@@ -0,0 +1,400 @@
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Sets.union;
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.TABLES;
+import static io.prestosql.metadata.MetadataListing.listSchemas;
+import static io.prestosql.metadata.MetadataListing.listTableColumns;
+import static io.prestosql.metadata.MetadataListing.listTablePrivileges;
+import static io.prestosql.metadata.MetadataListing.listTables;
+import static io.prestosql.metadata.MetadataListing.listViews;
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static io.prestosql.spi.type.TypeUtils.writeNativeValue;
+import static java.util.Objects.requireNonNull;
+
+public class InformationSchemaPageSource
+        implements ConnectorPageSource
+{
+    private enum InformationSchemaTableType
+    {
+        TABLE_COLUMNS,
+        TABLE_TABLES,
+        TABLE_VIEWS,
+        TABLE_SCHEMATA,
+        TABLE_TABLE_PRIVILEGES,
+        TABLE_ROLES,
+        TABLE_APPLICABLE_ROLES,
+        TABLE_ENABLED_ROLES
+    }
+
+    private final Session session;
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+    private final String catalogName;
+    private final Optional<Iterator<QualifiedTablePrefix>> tablePrefixIterator;
+
+    private final InformationSchemaTableType tableType;
+
+    private final List<Type> columnTypes;
+    private final Queue<Page> rawPages = new ArrayDeque<>();
+    private final PageBuilder rawPageBuilder;
+    private final List<Integer> outputPageChannels;
+
+    private long completedBytes;
+    private long memoryUsageBytes;
+    private boolean closed;
+
+    public InformationSchemaPageSource(
+            Session session,
+            Metadata metadata,
+            AccessControl accessControl,
+            String catalogName,
+            SchemaTableName schemaTableName,
+            Iterator<QualifiedTablePrefix> tablePrefixIterator,
+            List<ColumnHandle> outputColumns)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        requireNonNull(schemaTableName, "schemaTableName is null");
+
+        tableType = getTableTypeBySchemaTableName(schemaTableName);
+
+        // Tables "schemata", "roles", "applicable_roles", "enabled_roles" are catalog-wise and don't need a set of prefixes.
+        // The implementation in InformationSchemaMetadata::applyFilter indicates that a set of prefixes may still be built
+        // for such kind of tables. In this case, we don't use them.
```

Comment:
IMO this comment would be better expressed in the code as `checkState`, not `Optional`

Reply:
> I think we should get pages for these tables in the same place like for others. I don't see a reason why it is handled differently.
> 
> For example in case that these records were not actually needed there was no need to create them.
> 
> Also it does not match your commit message description. You described that you are loading data lazily, but here some data is loaded eagerly.

The reason is to simplify code in `isFinished`, otherwise I will have to keep a field variable indicating if these tables have been built. What's your opinion?

Reply:
Keeping a field is not an issue for me. Loading things eagerly could be a real production issue, let's avoid that if possible.

Reply:
I would simply:
```
checkState((tableType == InformationSchemaTableType.TABLE_SCHEMATA
                || tableType == InformationSchemaTableType.TABLE_ROLES
                || tableType == InformationSchemaTableType.TABLE_APPLICABLE_ROLES
                || tableType == InformationSchemaTableType.TABLE_ENABLED_ROLES) && !tablePrefixIterator.hasNext(), "");
```

and moved data generation for these tables to `getNextPage()`

Reply:
But it is not true that `tablePrefixIterator.hasNext()` is `false`. According to the original `InformationSchemaMetadata::getTableHandle` and `InformationSchemaMetadata::applyFilter`, for such tables, `prefixes` can be `defaultPrefixes()` and a set of schema prefixes.

Reply:
Right. So maybe you should have `if` in `getNextPage()` and simply don't use prefixes there. Like it was.